core/services/keystore/solana.go

```diff
@@ -0,0 +1,153 @@
+package keystore
+
+import (
+	"fmt"
+
+	"github.com/pkg/errors"
+
+	"github.com/smartcontractkit/chainlink/core/services/keystore/keys/solkey"
+	"github.com/smartcontractkit/chainlink/core/services/pg"
+)
+
+type Solana interface {
+	Get(id string) (solkey.Key, error)
+	GetAll() ([]solkey.Key, error)
+	Create() (solkey.Key, error)
+	Add(key solkey.Key) error
+	Delete(id string) (solkey.Key, error)
+	Import(keyJSON []byte, password string) (solkey.Key, error)
+	Export(id string, password string) ([]byte, error)
+	EnsureKey() (solkey.Key, bool, error)
+}
+
+type solana struct {
+	*keyManager
+}
+
+var _ Solana = &solana{}
+
+func newSolanaKeyStore(km *keyManager) *solana {
+	return &solana{
+		km,
+	}
+}
+
+func (ks *solana) Get(id string) (solkey.Key, error) {
+	ks.lock.RLock()
+	defer ks.lock.RUnlock()
+	if ks.isLocked() {
+		return solkey.Key{}, ErrLocked
+	}
+	return ks.getByID(id)
+}
+
+func (ks *solana) GetAll() (keys []solkey.Key, _ error) {
+	ks.lock.RLock()
+	defer ks.lock.RUnlock()
+	if ks.isLocked() {
+		return nil, ErrLocked
+	}
+	for _, key := range ks.keyRing.Solana {
+		keys = append(keys, key)
+	}
+	return keys, nil
+}
+
+func (ks *solana) Create() (solkey.Key, error) {
+	ks.lock.Lock()
+	defer ks.lock.Unlock()
+	if ks.isLocked() {
+		return solkey.Key{}, ErrLocked
+	}
+	key, err := solkey.New()
+	if err != nil {
+		return solkey.Key{}, err
+	}
+	return key, ks.safeAddKey(key)
+}
+
+func (ks *solana) Add(key solkey.Key) error {
+	ks.lock.Lock()
+	defer ks.lock.Unlock()
+	if ks.isLocked() {
+		return ErrLocked
+	}
+	if _, found := ks.keyRing.Solana[key.ID()]; found {
+		return fmt.Errorf("key with ID %s already exists", key.ID())
+	}
+	return ks.safeAddKey(key)
+}
+
+func (ks *solana) Delete(id string) (solkey.Key, error) {
+	ks.lock.Lock()
+	defer ks.lock.Unlock()
+	if ks.isLocked() {
+		return solkey.Key{}, ErrLocked
+	}
+	key, err := ks.getByID(id)
+	if err != nil {
+		return solkey.Key{}, err
+	}
+	err = ks.safeRemoveKey(key, func(tx pg.Queryer) error {
+		_, err2 := tx.Exec(`DELETE FROM p2p_peers WHERE peer_id = $1`, key.ID())
```

Comment:
Is this correct?

Reply:
Looks like a copypasta, be sure to check the rest of the file carefully just in case there are more mistakes.

Reply:
yeah looks wrong to me, fixed